### PR TITLE
feat: GSPO

### PR DESCRIPTION
**Pull request by pjin-nvidia**

### CODE REVIEW

File: nemo_rl/algorithms/loss_functions.py [hunk 1]
```diff
@@ -37,6 +37,10 @@ class ClippedPGLossConfig(TypedDict):
     use_on_policy_kl_approximation: bool
     use_importance_sampling_correction: bool
     token_level_loss: bool
+    # If True, apply the off-policy importance-sampling correction at the
+    # sequence level (one weight per generated sample).  If False (default)
+    # correction is applied at the token level as in the original GRPO paper.
+    sequence_level_importance_sampling: bool
```
Comment (ashors1) [inline nemo_rl/algorithms/loss_functions.py]: This naming seems a bit misleading to me because it seems to imply that we need `use_importance_sampling_correction` to be True in order for this to have an impact. But even when not using important sampling, `sequence_level_importance_sampling` impacts the loss computation. Do you think we can rename this to something more intuitive? Maybe something like `sequence_level_importance_ratios`?

Comment (pjin-nvidia) [reply nemo_rl/algorithms/loss_functions.py]: makes sense, going w/ `sequence_level_importance_ratios` !

**Comment by ashors1** [comment]
LGTM, thank you! Before merging, would it be possible to add some convergence plots to the PR description?

**Comment by ashors1** [comment]
Sorry, could you label the plots as well? It's not clear to me what is being shown in each

**Comment by pjin-nvidia** [comment]
added labels/titles to the example run convergence plots

File: examples/configs/gspo-deepscaler-1.5b-8K.yaml [hunk 2]
```diff
@@ -0,0 +1,7 @@
+# GRPO Algorithm Configuration
```
Comment (ashors1) [inline examples/configs/gspo-deepscaler-1.5b-8K.yaml]: Could we move this to `recipes/llm`? Also, we'll need to add a corresponding CI test. You can use this commit as guidance: https://github.com/NVIDIA-NeMo/RL/commit/d45ff3f7041937738a0247ee7efc65cf70079948. We need to add `tests/test_suites/llm/gspo-deepscaler-1.5b-8K.sh` and update `tests/test_suites/nightly.txt` to include the new test.

File: nemo_rl/algorithms/loss_functions.py [hunk 3]
```diff
@@ -229,11 +253,29 @@ def __call__(
...
+            # importance weight w_i = exp(Σ_t (log π_actor − log π_behaviour))
+            seq_lp_diff = ((prev_logprobs - generation_logprobs) * mask).sum(dim=-1)
+            actor_importance_weights = torch.exp(seq_lp_diff).detach()
+            actor_importance_weights = torch.nan_to_num(
+                actor_importance_weights, nan=0.0, posinf=0.0, neginf=0.0
+            )
+            # Broadcast to token dimension so we can reuse existing reduction
+            actor_importance_weights_expanded = actor_importance_weights.unsqueeze(-1)
+        else:
+            # Token-level correction (original GRPO implementation)
```
Comment (SahilJain314) [inline nemo_rl/algorithms/loss_functions.py]:
```suggestion
            # Token-level correction
```

Comment (SahilJain314) [reply nemo_rl/algorithms/loss_functions.py]: This correction isn't a part of 'vanilla GRPO' so I removed that part of the comment. (We explain it in the doc)

Comment (pjin-nvidia) [reply nemo_rl/algorithms/loss_functions.py]: thanks! removed that part of the comment

File: examples/configs/gspo-deepscaler-1.5b-8K.yaml (hunk 2 above)
Comment (pjin-nvidia) [reply examples/configs/gspo-deepscaler-1.5b-8K.yaml]: added the GSPO deepscaler run as a nightly test! (same config corresponding to the wandb plots in the OP)

**Comment by ashors1** [comment]
@terrykong if the PR looks good to you, can we merge?

File: nemo_rl/algorithms/loss_functions.py [hunk 4]
```diff
@@ -60,6 +65,7 @@ class ClippedPGLossFn(LossFunction):
     - PPO (Clipped) - https://arxiv.org/abs/1707.06347
     - GRPO - https://arxiv.org/abs/2402.03300
     - REINFORCE/RLOO (set disable_ppo_ratio = True and ignores ratio_clip_min/ratio_clip_max) - https://arxiv.org/abs/2402.14740
+    - GSPO (set sequence_level_importance_ratios = True and token_level_loss = False) - https://arxiv.org/abs/2507.18071
```
Comment (terrykong) [inline nemo_rl/algorithms/loss_functions.py:68]: could you add a unit test in https://github.com/NVIDIA-NeMo/RL/blob/df31c1bf428255e982303e69adaa19058c2a7970/tests/unit/algorithms/test_loss_functions.py#L4 that exercises this path?

**Review by ashors1 — DISMISSED** [review]

**Review by ashors1 — APPROVED** [review]
LGTM, thanks for the unit tests!
